src/Bicep.Core/Emit/ExpressionEmitter.cs

```diff
@@ -25,15 +24,18 @@ public class ExpressionEmitter
             SingleStringHandling = ExpressionSerializerSingleStringHandling.SerializeAsString
         });
 
-        private readonly JsonTextWriter writer;
+        private readonly PositionTrackingJsonTextWriter writer;
         private readonly EmitterContext context;
         private readonly ExpressionConverter converter;
 
-        public ExpressionEmitter(JsonTextWriter writer, EmitterContext context)
+        public readonly IDictionary<string, IDictionary<int, IList<(int start, int end)>>>? rawSourceMap;
```

Comment:
Please could we keep the state mutation out of this class and just have a dedicated class for managing this `Dictionary`?

Reply:
So I added this functionality to the position tracking writer but kept the class still implementing JsonTextWriter. That left other places still calling JsonTextWriter methods directly--which means a little more potential for maintainability of the source map since the new methods have to be explicitly used instead of the JsonTextWriter methods. However, pushing all interaction through the restricted interface would cause a huge amount of code churn with no functional change since most of these method calls are without an associated IPositionable object to associate with bicep lines.

I thought of maybe renaming the methods something like `WriteObjectWithTracking()` or something like that to indicate it does create mapping for the source map. Thoughts?